### PR TITLE
AppleTLS support

### PR DESCRIPTION
**Pull request by chamons**
- This PR ports AppleTLS support from https://github.com/xamarin/xamarin-macios/tree/master/src/Security to mono.
- On Xamarin.iOS and Xamarin.Mac builds we now only use the AppleTLS provider.
- On net_4_x, we default to AppleTLS when we're on the mac platform, but allow switching to other configured providers.

You can see a quick example of that here: https://gist.github.com/chamons/fbde13f07b1d767b0305744019a056f2

As a vast majority of the changes are porting from xamarin-macios with changes, I've created separate diffs showing the changes:

AppleCertificateHelper.cs - https://gist.github.com/596602e0747a88a46976b767877857f2
AppleTlsContext.cs - https://gist.github.com/50419f72f634be20805929d1a3d18764
AppleTlsProvider.cs - https://gist.github.com/be89c0f82e3b00b7cbc0638eb755506f
AppleTlsStream.cs - https://gist.github.com/4f6f561cc85c0ec796c67c2d0b40d45b
Certificate.cs - https://gist.github.com/5557a58f9678d787fb3c8bd1ac3af542
Enums.cs - https://gist.github.com/60cc68bc49e6b30414606b7d7173e380
ImportExport.cs - https://gist.github.com/95a8f244103674065774deb310cb1508
Items.cs - https://gist.github.com/chamons/09539202e1422aa8563118067a6389c2
Policy.cs - https://gist.github.com/87aba08b94e1ace8925f1b42f271d079
SecureTransport.cs - https://gist.github.com/905b70cc36dc38d22d882a5ba06b69f3
SslConnection.cs - https://gist.github.com/71d58eb7955eabfcdc7ffb3781f81e01
Trust.cs - https://gist.github.com/770693f0ec493cd89b2b61edeb239358

A majority of the changes while porting were in the following categories:

- Namespace and Using changes
- Porting from NS based APIs to CF based ones, which can be p/invoked directly.
- Removing some uses of System.Linq, which can't be used inside System.
- Manually dlsym'ing constants that are needed instead of depending on xamarin-macios "magic" infrastructure.

There are two feature flags for profiles/makefiles:

MONO_FEATURE_APPLETLS - Which enables building the AppleTLS invokes at all
ONLY_APPLETLS - Which switches out the default MonoTlsProviderFactory with one that only returns a wrapped AppleTLS provider.

### CODE REVIEW

**Comment by chamons** [comment]
@spouliot / @rolf - Could you please review my manual conversions from NS to CF APIs and dlsym changes to make sure they are safe?

File: mcs/class/System/System.Net/MacProxy.cs [hunk 1]
```diff
@@ -1066,4 +1273,90 @@ public static IWebProxy GetDefaultProxy ()
...
+			}
+		}
+
+		public void Dispose ()
+		{
+			Dispose (true);
+			GC.SuppressFinalize (this);
+		}
+
+#if XAMCORE_2_0
```
Comment (marek-safar) [inline mcs/class/System/System.Net/MacProxy.cs]: Why do you need this?

File: mcs/class/System/mobile_System.dll.sources [hunk 2]
```diff
@@ -1,2 +1,16 @@
 #include common.sources
 System.IO/FileSystemWatcher_mobile.cs
+
+Mono.AppleTls/AppleCertificateHelper.cs
```
Comment (marek-safar) [inline mcs/class/System/mobile_System.dll.sources]: This is duplicate, it's already in common

File: mcs/class/corlib/xammac_net_4_5_corlib.dll.sources [hunk 3]
```diff
@@ -0,0 +1 @@
+#include corlib.dll.sources
```
Comment (marek-safar) [inline mcs/class/corlib/xammac_net_4_5_corlib.dll.sources]: This file is not needed

File: mcs/class/corlib/corlib.dll.sources [hunk 4]
```diff
@@ -1661,3 +1661,7 @@ corefx/SR.cs
 ../../../external/corefx/src/System.Collections.Concurrent/src/System/Collections/Concurrent/OrderablePartitioner.cs
 ../../../external/corefx/src/System.Collections.Concurrent/src/System/Collections/Concurrent/Partitioner.cs
 ../../../external/corefx/src/System.Collections.Concurrent/src/System/Collections/Concurrent/PartitionerStatic.cs
+
+System.Security.Cryptography.X509Certificates/X509CertificateImplApple.cs
```
Comment (marek-safar) [inline mcs/class/corlib/corlib.dll.sources]: Can we move them to System assembly

File: mcs/class/System/System.Net/MacProxy.cs [hunk 5]
```diff
@@ -326,15 +391,144 @@ public override string ToString ()
...
+		}
+
+	}
+
+	
+	internal class CFData : CFObject
+	{
+		public CFData (IntPtr handle, bool own) : base (handle, own) { }
+	
+		[DllImport ("/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation")]
```
Comment (marek-safar) [inline mcs/class/System/System.Net/MacProxy.cs]: We already have constant for it, please use it everywhere

File: mcs/class/System/Test/System.Net.Security/SslStreamTest.cs [hunk 6]
```diff
@@ -59,6 +59,7 @@ public class SslStreamTest {
 	}
 
 	[Test] //bug https://bugzilla.novell.com/show_bug.cgi?id=457120
+	[Category ("MacNotWorking")] // Works but launches a prompt on 10.12 that will fail if you don't click in a few seconds
```
Comment (marek-safar) [inline mcs/class/System/Test/System.Net.Security/SslStreamTest.cs]: This is wrong, it's too general e.g. watch can still be tested

File: mcs/class/System/Mono.AppleTls/INativeObject.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+using System;
+
+namespace ObjCRuntime {
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/INativeObject.cs]: Use Mono.ObjCRuntime and move it to correct location

File: mcs/class/System/Mono.AppleTls/SecureTransport.cs [hunk 8]
```diff
@@ -0,0 +1,254 @@
+#if MONO_FEATURE_APPLETLS
+// Copyright 2014 Xamarin Inc. All rights reserved.
+
+using System;
+using ObjCRuntime;
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/SecureTransport.cs]: Why these usings

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs [hunk 9]
```diff
@@ -1,24 +1,147 @@
...
+#if SECURITY_DEP
+			return (IMonoTlsProvider)GetTlsProvider ();
+#else
+			throw new NotSupportedException ("TLS Support not available.");
+#endif
+		}
+		
+		#endregion
+
+		static object locker = new object ();
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: should be under #if

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs [hunk 10]
```diff
@@ -1,24 +1,147 @@
...
+		{
+			return true;
+		}
+
+		internal static MSI.MonoTlsProvider GetProvider (string name)
+		{
+			return GetTlsProvider ().Provider;
+		}
+
+		internal static bool IsInitialized => true;
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: What is purpose of this property?

File: mcs/class/System/Mono.AppleTls/SslConnection.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Copyright 2014 Xamarin Inc.
+//
+
+using System;
+using System.IO;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+
+using ObjCRuntime;
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/SslConnection.cs]: Too many unused usings

File: mcs/class/System/Mono.AppleTls/Enums.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
+#if MONO_FEATURE_APPLETLS
+// Copyright 2011-2015 Xamarin Inc. All rights reserved.
+using System;
+
+using ObjCRuntime;
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/Enums.cs]: Redundant usings

File: mcs/class/System/Mono.AppleTls/Items.cs [hunk 13]
```diff
@@ -0,0 +1,258 @@
+#if SECURITY_DEP && MONO_FEATURE_APPLETLS
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/Items.cs]: You don't need SECURITY_DEP check

File: mcs/class/System/Mono.AppleTls/Trust.cs [hunk 14]
```diff
@@ -0,0 +1,186 @@
+#if SECURITY_DEP && MONO_FEATURE_APPLETLS
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/Trust.cs]: Extra SECURITY_DEP

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 15]
```diff
@@ -0,0 +1,931 @@
...
+					throw new ObjectDisposedException ("AppleTlsContext");
+				return context;
+			}
+		}
+
+		public override bool HasContext {
+			get { return !disposed && context != IntPtr.Zero; }
+		}
+
+		[SD.Conditional ("MARTIN_DEBUG")]
```
Comment (marek-safar) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: Could we replace this with something more meaningful

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs [hunk 16]
```diff
@@ -1,24 +1,147 @@
...
+#endif
+using System.Security.Cryptography.X509Certificates;
+#endif
+
+using System;
+using System.Net;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+#if SECURITY_DEP
+using Mono.AppleTls;
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: Use the #if scope above

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs [hunk 17]
```diff
@@ -1,24 +1,147 @@
...
+		 * reflection or from friend assemblies.
+		 * 
+		 * @IMonoTlsProvider is defined as empty interface outside 'SECURITY_DEP', so we don't need
+		 * this conditional here.
+		 */
+
+		internal static IMonoTlsProvider GetProviderInternal ()
+		{
+#if SECURITY_DEP
+			return (IMonoTlsProvider)GetTlsProvider ();
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: Why the cast?

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs [hunk 18]
```diff
@@ -1,24 +1,147 @@
...
+
+		internal static void Initialize ()
 		{
-			MSI.MonoTlsProvider provider = null;
-			if (MSI.MonoTlsProviderFactory._PrivateFactoryDelegate != null)
-				provider = MSI.MonoTlsProviderFactory._PrivateFactoryDelegate ();
-			return provider;
 		}
+
+		internal static void Initialize (string provider)
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: Is this used anywhere?

**Comment by rolfbjarne** [comment]
Chris, I think this would be much easier to review if you did two commits like this:

1. First move all the needed files to the mono repo (without changing them), and commit these. This commit should not be reviewed.
2. Then make all the required changes, and make another commit.

This is easier to review, because now we can more easily skip the first commit (which are just file moves, and not your code), and then only review the second commit (I can see your separate file diffs/gists, but I can't comment on those).

It should not be difficult to do this (create a new branch off master, move the files, commit 1, copy the  files from this branch, commit 2), but it would end up with a different PR.

That would also make it clear if Marek's comments are related to your changes or not.

File: mcs/class/System/System.Net/MacProxy.cs (hunk 1 above)
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: It was copy/paste from xamarin-macios. Will fix.

File: mcs/class/System/System.Net/MacProxy.cs (hunk 5 above)
Comment (chamons) [inline mcs/class/System/System.Net/MacProxy.cs]: Will fix.

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs (hunk 10 above)
Comment (chamons) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: To match the exact API in MonoTlsProviderFactory.cs

File: mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs (hunk 18 above)
Comment (chamons) [inline mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.Apple.cs]: I'm matching the API from MonoTlsProviderFactory.cs

File: mcs/class/corlib/corlib.dll.sources (hunk 4 above)
Comment (chamons) [inline mcs/class/corlib/corlib.dll.sources]: No, we're already doing this in monotouch_corlib.dll.sources and the non-mobile IntPtr version is here as well (refactoring this is outside of scope).

**Pull request closed without merging**